rfcs/2021-08-13-8025-internal-tracing.md

```diff
@@ -0,0 +1,254 @@
+long-lived spans. It's possible that some components may not have a clear
+concept of a single "iteration" around which we can wrap a base span, and we'll
+have a hard time instrumenting them in a way that is both logical and
+practically useful.
+
+### What kind on ongoing burden does this place on the team?
+
+The ongoing maintenance burden should be quite minimal. The extent of code
+changes would be the addition of some small span instantiations and
+instrumentation, but nothing that would require threading through existing
+interfaces or reorganizing any abstractions.
+
+## Prior Art
+
+There is very little prior art available for tracing streaming system, but we
+are largely just using existing off-the-shelf components. We've not opted to
+invent anything particularly new for Vector.
+
+## Alternatives
+
+### What other approaches have been considered and why did you not choose them?
+
+The primary alternative implementation that's been discussed is to have spans be
+event-focused rather than component-focused. This would involve stashing a root
+span instance in a Vector event's metadata, and creating child spans to
+represent each stage of its processing. This would provide an end-to-end view of
+an event's journey through Vector, showing where the event itself spent time as
+opposed to where each component is spending time. While this would produce some
+interesting data, it would be significantly more complex and manual to
+implement, and it's not clear that the data would actually be more useful in
+addressing performance issues.
+
+Another alternative would be to avoid the coupling between tracing spans and
+logs, and use a library other than `tracing` for this functionality. This may
+provide some benefit in reducing confusion about log levels, accumulating events
+on spans, etc, but has the downside of requiring the introduction of a second
+system for instrumentation. It would also be rather confusing to use the
+`tracing` crate, but not for tracing, especially since it is one of the most
+widely used and well-engineered crates for this purpose.
+
+### How about not doing this at all?
+
+As discussed in the Rationale section, not doing this at all would leave us with
+a meaningful gap in our observability. There would be no immediate dire
+consequences, but we would expect to continue sinking more engineering time into
+debugging customer issues than we would if this data were easily available.
+
+## Outstanding Questions
+
+- What is an acceptable performance overhead for enabling tracing in user environments?
+- Are component-focused spans acceptable, or are there cases where event-focused spans would be clearly superior?
+- Are we comfortable with the proposed methods of avoiding the previous memory growth issue?
+
+## Plan Of Attack
+
+Incremental steps to execute this change. These will be converted to issues
+after the RFC is approved:
+
+- [ ] Adjust existing spike-level implementation to fully match what's described here
```

Comment:
I have had some questions about tracing's runtime impact. If we could contrive a feature flag that let us turn off tracing that'd be valuable for getting good performance impact numbers.

Reply:
AFAIK, `tracing` is designed such that events and spans with no interested subscribers are never constructed. This effectively means that by turning the log level up to `CRITICAL` or similar, very little should actually happen at `span` and `event` callsites. That's obviously not the same as turning it off entirely, but in practice it's intended to be very close.

Reply:
Oh yeah, that'd be an interesting experiment and might let us cheaply get at what a "no span" vector looks like, with regard to throughput.